onnxruntime/test/global_thread_pools/test_inference.cc

```diff
@@ -0,0 +1,222 @@
+  std::vector<Ort::Value> ort_inputs;
+  std::vector<const char*> input_names;
+  for (size_t i = 0; i < inputs.size(); i++) {
+    input_names.emplace_back(inputs[i].name);
+    ort_inputs.emplace_back(Ort::Value::CreateTensor<float>(allocator->Info(allocator), const_cast<float*>(inputs[i].values.data()), inputs[i].values.size(), inputs[i].dims.data(), inputs[i].dims.size()));
+  }
+
+  std::vector<Ort::Value> ort_outputs;
+  if (output_tensor)
+    session_object.Run(Ort::RunOptions{nullptr}, input_names.data(), ort_inputs.data(), ort_inputs.size(), &output_name, output_tensor, 1);
+  else {
+    ort_outputs = session_object.Run(Ort::RunOptions{nullptr}, input_names.data(), ort_inputs.data(), ort_inputs.size(), &output_name, 1);
+    ASSERT_EQ(ort_outputs.size(), 1u);
+    output_tensor = &ort_outputs[0];
+  }
+
+  auto type_info = output_tensor->GetTensorTypeAndShapeInfo();
+  ASSERT_EQ(type_info.GetShape(), dims_y);
+  //size_t total_len = type_info.GetElementCount();
+  ASSERT_EQ(values_y.size(), static_cast<size_t>(5));
+
+  OutT* f = output_tensor->GetTensorMutableData<OutT>();
+  for (size_t i = 0; i != static_cast<size_t>(5); ++i) {
+    ASSERT_TRUE(abs(values_y[i] - f[i]) < 1e-6);
+  }
+}
+
+template <typename T, typename OutT>
+static Ort::Session GetSessionObj(Ort::Env& env, T model_uri, int provider_type) {
+  Ort::SessionOptions session_options;
+  session_options.DisablePerSessionThreads();
+
+  if (provider_type == 1) {
+#ifdef USE_CUDA
+    Ort::ThrowOnError(OrtSessionOptionsAppendExecutionProvider_CUDA(session_options, 0));
+    std::cout << "Running simple inference with cuda provider" << std::endl;
+#else
+    return Ort::Session(nullptr);
+#endif
+  } else if (provider_type == 2) {
+#ifdef USE_DNNL
+    Ort::ThrowOnError(OrtSessionOptionsAppendExecutionProvider_Dnnl(session_options, 1));
+    std::cout << "Running simple inference with dnnl provider" << std::endl;
+#else
+    return Ort::Session(nullptr);
+#endif
+  } else if (provider_type == 3) {
+#ifdef USE_NUPHAR
+    Ort::ThrowOnError(OrtSessionOptionsAppendExecutionProvider_Nuphar(session_options, /*allow_unaligned_buffers*/ 1, ""));
+    std::cout << "Running simple inference with nuphar provider" << std::endl;
+#else
+    return Ort::Session(nullptr);
+#endif
+  } else {
+    std::cout << "Running simple inference with default provider" << std::endl;
+  }
+
+  // if session creation passes, model loads fine
+  return Ort::Session(env, model_uri, session_options);
+}
+
+template <typename T, typename OutT>
+static void TestInference(Ort::Session& session,
+                          const std::vector<Input>& inputs,
+                          const char* output_name,
+                          const std::vector<int64_t>& expected_dims_y,
+                          const std::vector<OutT>& expected_values_y) {
+  Ort::SessionOptions session_options;
+  session_options.DisablePerSessionThreads();
+  auto default_allocator = onnxruntime::make_unique<MockedOrtAllocator>();
```

Comment:
Are these 2 lines doing anything?

Reply:
They're preallocating the output tensor.

Reply:
Comment seems to have moved. I meant the 2 session_options related lines above.

---
In reply to: [393474760](https://github.com/microsoft/onnxruntime/pull/3177#discussion_r393474760) [](ancestors = 393474760)